Fix broken eunit test in changes_since_test_ test suite

jiangphcn opened the pull request:
<!-- Thank you for your contribution!
     
     Please file this form by replacing markdown commentary
     tags with the text. If section needs in no action - remove it.

     Also remember, that CouchDB uses the Review-Then-Commit (RTC) model 
     of code collaboration. Positive feedback provides by +1 from committers
     while negative by -1. The -1 also means veto and need to be addressed
     to find the consensus. Once there are no objections, PR could be merged.

     See: http://couchdb.apache.org/bylaws.html#decisions for more info -->

## Overview

<!-- Please give a short brief for the pull request,
     what problem it solves or how it makes things better. -->
Prior to the change in this PR, most test cases in two test suites `changes_since_test_` and `changes_index_test_` failed to execute. They were disabled. Considering different scope and historical reason, this PR is aimed to address failure from `changes_since_test_` test suites. The test suite `changes_index_test_` will be another story.

The change not only includes test code, but also includes minor change of couch_mrview_util.erl to satisfy required pattern matches.

## Testing recommendations
```
======================== EUnit ========================
module 'couch_mrview_changes_since_tests'
  changes_since tests
    couch_mrview_changes_since_tests:109: test_basic...ok
    couch_mrview_changes_since_tests:132: test_basic_since...ok
    couch_mrview_changes_since_tests:145: test_basic_count...ok
    couch_mrview_changes_since_tests:154: test_basic_count_since...ok
    couch_mrview_changes_since_tests:164: test_compact...ok
    couch_mrview_changes_since_tests:164: test_compact...ok
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
    [done in 0.666 s]
  changes_since_range tests
    couch_mrview_changes_since_tests:120: test_range...ok
    couch_mrview_changes_since_tests:141: test_range_since...ok
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
    [done in 0.236 s]
  changes_since_range_count tests
    couch_mrview_changes_since_tests:150: test_range_count...ok
    couch_mrview_changes_since_tests:159: test_range_count_since...ok
    couch_mrview_changes_since_tests:196: test_remove_key...ok
    couch_mrview_changes_since_tests:197: test_remove_key...ok
    couch_mrview_changes_since_tests:198: test_remove_key...ok
    couch_mrview_changes_since_tests:199: test_remove_key...ok
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
    [done in 0.423 s]
  [done in 3.173 s]
=======================================================
  All 14 tests passed.
==> rel (eunit)
=======================================================
```
Furthermore, the whole test suites from `couch_mrview` can execute successfully.
```
make check apps=couch_mrview
...
    couch_mrview_all_docs_tests:126: should_query_with_include_docs...ok
    couch_mrview_all_docs_tests:133: should_query_empty_views...ok
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
    [done in 0.537 s]
  [done in 0.635 s]
=======================================================
  All 129 tests passed.
==> rel (eunit)
```

<!-- Describe how we can test your changes.
     Does it provides any behaviour that the end users 
     could notice? -->

## JIRA issue number
COUCHDB-3360 
<!-- If this is a significant change, please file a JIRA issue at:
     https://issues.apache.org/jira/browse/COUCHDB
     and include the number here and in commit message(s)  -->

## Related Pull Requests
N/A
<!-- If your changes affects on multiple components in different 
     repositories please list here links to those pull requests.  -->

## Checklist

- [X] Code is written and works correctly;
- [X] Changes are covered by tests;
- [] Documentation reflects the changes;
- [X] I will not forget to update [rebar.config.script](https://github.com/apache/couchdb/blob/master/rebar.config.script)
      with the correct commit hash once this PR get merged.

eiri commented on src/couch_mrview/src/couch_mrview_test_util.erl:
```diff
@@ -53,7 +53,8 @@ ddoc(changes) ->
         {<<"options">>, {[
-            {<<"seq_indexed">>, true}
+            {<<"seq_indexed">>, true},
+            {<<"keyseq_indexed">>, true}
```
It is valid to have only seq btree enabled on a view, i.e. all tests should pass even with key-seq btree not built. If anything we should expand the tests to assert both variants.

Can you explain, please, why do you think changing a scope of `couch_mrview_changes_since_tests.erl` to test `key-seq` instead of `seq` is the right fix?

jiangphcn replied:
If query only contains "since", it should be enough to enable "seq_indexed". However, in some of test cases in this test suite, the query contains both "start_key" and "end_key".  This needs to enable "keyseq_indexed". Thus, the test can check `key_byseq_btree` instead of `seq_btree` to get expected results. I sorted out the test cases and only enable these two options when necessary.

jiangphcn commented on src/couch_mrview/src/couch_mrview_util.erl:
```diff
@@ -344,6 +344,9 @@ get_view_changes_count(View) ->
         {nil, #btree{}} ->
+            couch_btree:fold_reduce(KSBtree, CountFun, 0, []);
+        {#btree{}, #btree{}} ->
+            couch_btree:fold_reduce(SBtree, CountFun, 0, []),
```
Good suggestion, and already applied.

jiangphcn commented on src/couch_mrview/src/couch_mrview.erl:
```diff
@@ -708,5 +708,5 @@ make_view_changes_opts(StartSeq, Options, Args) ->
             couch_mrview_util:changes_key_opts(StartSeq, Args);
         false ->
-            [[{start_key, {StartSeq+1, <<>>}}] ++ Options]
+            [[{start_key, {StartSeq, <<>>}}] ++ Options]
```
Eric, thanks for your comments. 

This is based on current understanding of query with only since.
```
test_basic_since(_, Db) ->
    Result = run_query(Db, 5, []),
    Expect = {ok, [
                {{6, 4, <<"4">>}, 4},
                {{7, 5, <<"5">>}, 5},
                {{8, 6, <<"6">>}, 6},
                {{9, 7, <<"7">>}, 7},
                {{10, 8, <<"8">>}, 8},
                {{11, 9, <<"9">>}, 9}
    ]},
    ?_assertEqual(Result, Expect).
```
If since is specified with 5, the expected returned document sequence should start from 6 instead of 7. Thus, there is change from [[{start_key, {StartSeq+1, <<>>}}] ++ Options] to [[{start_key, {StartSeq, <<>>}}] ++ Options]

Today, I test using one local cluster, and can find some test result related to `startkey` and `endkey`.
```
jiangphs-MacBook-Pro:geotest jiangph$ curl -u foo:bar -X GET 'http://localhost:15984/sdb1/_design/bar/_view/baz?startkey=3&endkey=4'|jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   101    0   101    0     0   8824      0 --:--:-- --:--:-- --:--:--  9181
{
  "total_rows": 6,
  "offset": 2,
  "rows": [
    {
      "id": "3",
      "key": 3,
      "value": 3
    },
    {
      "id": "4",
      "key": 4,
      "value": 4
    }
  ]
}
```

However, for `since`, I still work from eunit test case.

jiangphcn replied:
Yes, this makes sense. This is really subtle bug. Need to take care of different situations. We can use a different PR to address this.

jiangphcn commented:
Hi @eiri Thanks so much for reviewing codes and providing suggestion. I almost addressed your comments. However, regarding whether to set StartSeq with or without plus 1 for query only with `since`, we still need some discussion/test.

eiri commented on src/couch_mrview/src/couch_mrview_test_util.erl:
```diff
@@ -72,6 +72,53 @@ ddoc(changes) ->
             ]}}
         ]}}
     ]});
+ddoc({changes, keyseq_indexed}) ->
```
This is rather repetitive, to be honest, when we only want to change options. How changing original `doc(changes)` to this:
```
ddoc({changes, Opts}) ->
    ViewOpts = case Opts of
        seq_indexed ->
            [{<<"seq_indexed">>, true}];
        keyseq_indexed ->
            [{<<"keyseq_indexed">>, true}];
        seq_indexed_keyseq_indexed ->
            [
                {<<"seq_indexed">>, true},
                {<<"keyseq_indexed">>, true}
            ]
    end,
    couch_doc:from_json_obj({[
        {<<"_id">>, <<"_design/bar">>},
        {<<"options">>, {ViewOpts}},
        {<<"views">>, {[
            {<<"baz">>, {[
                {<<"map">>, <<"function(doc) {emit(doc.val, doc.val);}">>}
            ]}}
        ]}}
    ]});
```

Same result, way less typing and it's a bit clearer on what's going on.
Also we don't need `bing` and `zing` since we don't use them in the tests.

eiri commented on src/couch_mrview/src/couch_mrview.erl:
```diff
@@ -267,7 +267,7 @@ view_changes_since(View, StartSeq, Fun, Opts0, Acc) ->
         Fun(KV, Acc1)
     end,
-    Opts = [{start_key, {StartSeq + 1, <<>>}}] ++ Opts0,
+    Opts = [{start_key, {StartSeq, <<>>}}] ++ Opts0,
```
Revert this change. See the explanation below.

eiri commented on src/couch_mrview/test/couch_mrview_changes_since_tests.erl:
```diff
@@ -92,37 +129,37 @@ test_basic_since(Db) ->
-test_compact(Db) ->
+test_compact(_, Db) ->
     Result = couch_mrview:compact(Db, <<"_design/bar">>),
     ?_assertEqual(Result, ok),
```
There are for warnings for all those tests that looks like "Warning: a term is constructed, but never used". I'll explain what's wrong taking this line as an example.

In erlang unit tests all the macros starting with underscore not executed immediately, but instead return function wrapping the assertion. Function `test_compact` here is what called generator and it supposed to return either function to execute as a test or list of such function. We are returning test function generated with last line `?_assertEqual(Count, 10)`, so it got tested.

But call `?_assertEqual(Result, ok)` is not actually testing anything, it returns function that we are throwing away and this is what the warning about - we are constructing a function, but never use it. To fix it we need to return list of tests, like this: `[?_assertEqual(Result, ok), ?_assertEqual(Count, 10)]`

There are four such places in this test module, please fix all of them.

jiangphcn commented:
Thanks so much @eiri for your review and suggestions. I have addressed all of your current comments. You also mentioned that we want to have another PR to introduce "customer less". So don't change in this PR. 

I run test suites below and they can pass. Finally, thanks again! 

make eunit apps=couch_mrview suites=couch_mrview_changes_since_tests
```
======================== EUnit ========================
module 'couch_mrview_changes_since_tests'
  changes_since tests
    couch_mrview_changes_since_tests:109: test_basic...ok
    couch_mrview_changes_since_tests:132: test_basic_since...ok
    couch_mrview_changes_since_tests:145: test_basic_count...ok
    couch_mrview_changes_since_tests:154: test_basic_count_since...ok
    couch_mrview_changes_since_tests:164: test_compact...ok
    couch_mrview_changes_since_tests:164: test_compact...ok
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
    [done in 0.666 s]
  changes_since_range tests
    couch_mrview_changes_since_tests:120: test_range...ok
    couch_mrview_changes_since_tests:141: test_range_since...ok
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
    [done in 0.236 s]
  changes_since_range_count tests
    couch_mrview_changes_since_tests:150: test_range_count...ok
    couch_mrview_changes_since_tests:159: test_range_count_since...ok
    couch_mrview_changes_since_tests:196: test_remove_key...ok
    couch_mrview_changes_since_tests:197: test_remove_key...ok
    couch_mrview_changes_since_tests:198: test_remove_key...ok
    couch_mrview_changes_since_tests:199: test_remove_key...ok
[os_mon] cpu supervisor port (cpu_sup): Erlang has closed
    [done in 0.423 s]
  [done in 3.173 s]
=======================================================
  All 14 tests passed.
==> rel (eunit)
```
make eunit apps=couch_mrview
```
    [done in 0.589 s]
  [done in 0.734 s]
=======================================================
  All 129 tests passed.
==> rel (eunit)
==> couchdb (eunit)
```

jiangphcn commented:
@eiri thanks again for review and approval. Because I didn't have write access to  `https://github.com/apache/couchdb`, so currently I am working on `https://github.com/cloudant/couchdb`. I rebased from `https://github.com/cloudant/couchdb` and  squash commits. Right now, I can see `All Checks have passed`. Would you please help check and merge to  `https://github.com/apache/couchdb` if applicable? Thanks.
